### PR TITLE
Unifall episode 2: apply, the comeback

### PR DESCRIPTION
**Pull request by mattam82**
This set of commits ports apply to the evar_conv unification and evar-based clauses, leaving almost everything else unchanged (it still impacts reflexivity and eauto's exact which use apply or clenv's unify primitive). Right now I'm just curious about potential compatibility issues on the CI. The ones I found in the stdlib are either:

- apply/eapply discrepancies, apply was allowing itself to produce dependent subgoals sometimes
- different instantiation of holes in apply in non-linear cases: the new apply prefers the first instance it finds (left-to-right traversal of the term)
- different occurrence selection in higher-order unification, due to different heuristics being used. This is rare as often the user provides enough info (e.g. a predicate) to disambiguate.

Fixes / closes #????

- [ ] Added / updated **test-suite**.

- [ ] Added **changelog**.
- [ ] Added / updated **documentation**.
  
  - [ ] Documented any new / changed **user messages**.

- [ ] Opened **overlay** pull requests.

### CODE REVIEW

**Comment by JasonGross** [comment]
@coqbot ci minimize

**Comment by Alizter** [comment]
@JasonGross Did coqbot just race with itself causing some of the minimizer jobs to fail?

**Comment by JasonGross** [comment]
- ci-sf failed because https://gitlab.com/coq/coq/-/jobs/1991212828/artifacts/download is a 404, idk what's up here
- ci-itauto failed because https://gitlab.com/coq/coq/-/jobs/1991212857/artifacts/download is a 404
- ci-relation_algebra failed because https://gitlab.com/coq/coq/-/jobs/1991212883/artifacts/download is a 404
- ci-metacoq failed because https://gitlab.com/coq/coq/-/jobs/1991212867/artifacts/download is a 404
- [ci-hott failed to minimize](https://github.com/coq/coq/pull/15518#issuecomment-1017939088) because of an issue making universe names match in the error messages?
- [ci-rewriter succeeded](https://github.com/coq/coq/pull/15518#issuecomment-1017939694)
- [ci-bedrock2 succeeded](https://github.com/coq/coq/pull/15518#issuecomment-1017939985)
- [ci-quickchick succeeded](https://github.com/coq/coq/pull/15518#issuecomment-1017941116)
- (a bunch of others succeded)
- [ci-category_theory failed](https://github.com/coq/coq/pull/15518#issuecomment-1017944365) because `git clone https://github.com/coq/coq.git` gave "fatal: unable to access 'https://github.com/coq/coq.git/': GnuTLS recv error (-54): Error in the pull function."
- (the rest so far succeeded)

@coqbot ci minimize ci-category_theory

**Comment by JasonGross** [comment]
Idk why it posted a "ready to minimize" a second time, but maybe this'll work now? @coqbot ci minimize ci-category_theory ci-sf ci-itauto ci-relation_algebra ci-metacoq

File: compat_unif [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  -> ?R does not become a subgoal, ?R dependent
+
+  No difference? ?MR is an initial _meta_ of the lemma, with ?R' :=
+  relprod ?R' ?S with ?MR := ?R', an evar.  ?Me is an initial meta of
+  the lemma with ?X := equiv ?e'.  ?Me := e'
+
+  Here ?R is a meta instantiated by an evar and unify_resolve succeeds
+  without seeing that evar is left around, as clenv_pose_dependent_evars
+  was just looking at the meta substitution which contains no meta
+  without instance.
```
Comment (herbelin) [inline compat_unif:44]: I wonder whether "eapply could shelve subgoals before" is related to this comment of [#15521](https://github.com/coq/coq/issues/15521#issuecomment-1416848455) about a difference between `apply` and `rapply`?

**Pull request closed without merging**